crates/subspace-farmer/src/utils/farmer_piece_getter.rs

```diff
@@ -120,6 +158,46 @@ where
 
     /// Slow way to get piece using archival storage
     pub async fn get_piece_slow(&self, piece_index: PieceIndex) -> Option<Piece> {
+        let in_progress_piece_mutex = Arc::new(AsyncMutex::new(None));
```

Comment:
This code is almost identical to the previous function and it is complex enough to remove the duplication.

Reply:
I anticipated this comment. Yes, there are 3 places with almost identical, but slightly different code. However, de-duplicating it (I guess without macros that reduce readability) is challenging due to `MutexGuard`s that would require self-referrential types. I decided to not go that route, but feel free to play with it over the weekend and send a PR if you find a nicer way to handle this.

Reply:
Makes sense.